### DISCUSSION TITLE
Possible to refactor certain parts of API for streaming?

### DISCUSSION THREAD

**Discussion by LuminescentMoon** [quietvoid/dovi_tool#154] [Ideas]
I'm interested in using dovi_tool as a library to convert dual track profile 7.1 to 8.1. Is it possible, with possibly minor refactoring of dovi_tool, to perform said conversion with a stream-like API without storing entire video files in memory or touching the disk?

**Comment by quietvoid** [comment]
I don't understand what you're looking for.

`dovi_tool` is not a library, it is meant to be used as an executable.

You can already convert on the fly using ffmpeg, in the end an output file is still required.

**Comment by LuminescentMoon** [comment]
I understand but a large portion of the code required to perform the conversion is in the dovi_tool crate. Currently, it is not possible to convert on the fly without at least reading one file from disk as there can only be one input stream.

**Reply by quietvoid** [reply]
> I understand but a large portion of the code required to perform the conversion is in the dovi_tool crate.

It is in the `dolby_vision` crate, which is the core library.
You can easily convert input RPUs using that, and integrate your own I/O.


> Currently, it is not possible to convert on the fly without at least reading one file from disk as there can only be one input stream.

I still don't understand what your input is, the HEVC has to come from somewhere.
Please clarify.

**Comment by LuminescentMoon** [comment]
> It is in the dolby_vision crate, which is the core library.

There are a few hundred lines of code in muxer.rs involving knowledge of NALUs and ordering of BL/EL frames to mux a BL and an EL track.

> I still don't understand what your input is, the HEVC has to come from somewhere.

I would like to pipe in the BL and EL tracks from my code.

**Reply by quietvoid** [reply]
> I would like to pipe in the BL and EL tracks from my code.

Okay, that should be feasible if you pass in the `Read` for both. It would need a refactor.

I created the issue #155 for this.
